### PR TITLE
Add skipless demucs 2

### PR DESCRIPTION
**Pull request by mosem**

### CODE REVIEW

File: conf/experiment/demucs_skipless_1.yaml [hunk 1]
```diff
@@ -0,0 +1,49 @@
+experiment_name: demucs_skipless_1
```
Comment (Or-Tal) [inline conf/experiment/demucs_skipless_1.yaml:1]: add # @package experiment in the top line

File: conf/experiment/demucs_skipless_adversarial_1.yaml [hunk 2]
```diff
@@ -0,0 +1,58 @@
+experiment_name: demucs_skipless_adversarial_1
```
Comment (Or-Tal) [inline conf/experiment/demucs_skipless_adversarial_1.yaml:1]: add # @Package experiment in the top line

File: denoiser/models/autoencoder_composer.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    def forward(self, signal):
+        if self.skips:
+            latent, skips_signals = self.encoder(signal)
+            latent = self.attention_module(latent)
+            out = self.decoder(latent, skips_signals)
+        else:
+            latent = self.encoder(signal)
+            latent = self.attention_module(latent)
+            out = self.decoder(latent)
+        return out
```
Comment (Or-Tal) [inline denoiser/models/autoencoder_composer.py:30]: looks good

File: denoiser/models/autoencoder_composer.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    @capture_init
+    def __init__(self, encoder, attention_module, decoder, skips):
+        super().__init__()
+        self.encoder = encoder
+        self.attention_module = attention_module
+        self.decoder = decoder
+        self.skips = skips
+
+    def estimate_valid_length(self, input_length):
+        encoder_output_length = self.encoder.estimate_output_length(input_length)
```
Comment (Or-Tal) [inline denoiser/models/autoencoder_composer.py:16]: looks like a good practice to force this API on the modules. (estimate_output_length)

File: denoiser/models/demucs_encoder.py [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+            for encode in self.encoder:
+                x = encode(x)
+                skips_signals.append(x)
+
+            return std * x, skips_signals
+
+        else:
+            for encode in self.encoder:
+                x = encode(x)
+            return std * x
```
Comment (Or-Tal) [inline denoiser/models/demucs_encoder.py:139]: dividing by the std forces a normalization of the std to match a normal distribution (std=1) yet re-multiplying by std aims to un-normalize the output back to the same distribution. therefore multiplying by std should be done over the output of the ENTIRE network, i.e. in the decoder output.

File: denoiser/models/demucs_encoder.py [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+        elif self.resample == 4:
+            x = upsample2(x)
+            x = upsample2(x)
+        if self.skips:
+            skips_signals = []
+            for encode in self.encoder:
+                x = encode(x)
+                skips_signals.append(x)
+
+            return std * x, skips_signals
```
Comment (Or-Tal) [inline denoiser/models/demucs_encoder.py:134]: dividing by the std forces a normalization of the std to match a normal distribution (std=1) yet re-multiplying by std aims to un-normalize the output back to the same distribution. therefore multiplying by std should be done over the output of the ENTIRE network, i.e. in the decoder output.

File: test_denoiser.py [hunk 7]
```diff
@@ -12,28 +12,34 @@
...
+                'experiment.segment=2', 'experiment.stride=2', 'ddp=0',
+                'eval_every=1', 'epochs=1', f'hydra.run.dir={OUTPUT_DIR}', f'device={device}'],
     'caunet_adversarial': ['train.py', 'dset=valentini_dummy', 'experiment=caunet_adversarial_1',
-               'experiment.segment=2', 'experiment.stride=2', 'ddp=0',
-               'experiment.batch_size=1',
-               'eval_every=1', 'epochs=1', f'hydra.run.dir={OUTPUT_DIR}', f'device={device}'],
+                'experiment.segment=2', 'experiment.stride=2', 'ddp=0',
+                'experiment.batch_size=1',
+                'eval_every=1', 'epochs=1', f'hydra.run.dir={OUTPUT_DIR}', f'device={device}'],
     'demucs_hifi': ['train.py', 'dset=valentini_dummy', 'experiment=demucs_hifi',
```
Comment (Or-Tal) [inline test_denoiser.py:31]: I may have missed removing demucs_hifi from this section when I removed it from master, please remove this alongside fixing the above issues and merge into master without this segment

File: denoiser/models/demucs_decoder.py [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+        for idx in range(self.depth):
+            length = (length - 1) * self.stride + self.kernel_size
+        length = int(math.ceil(length / self.resample))
+        return int(length)
+
+    def forward(self, signal, skips=None):
+        if signal.dim() == 2:
+            signal = signal.unsqueeze(1)
+
+        if self.normalize:
```
Comment (Or-Tal) [inline denoiser/models/demucs_decoder.py:101]: this should be evalueated over the input to the entire network. I believe the best way to do so is to implement it in the wrapping model, such that normaliation is evaluated over the input and being applied on the input, and then re-multiply the output by std to get an un-normalized result.

**Review by Or-Tal — CHANGES_REQUESTED** [review]
Please look into my comments, I have addressed several important issues that you should handle before merging this branch. 
once fixed it could be merged. no need to re-inspect this one.

File: denoiser/models/autoencoder_composer.py (hunk 4 above)
Comment (mosem) [reply denoiser/models/autoencoder_composer.py:16]: I'll do this refactoring in another pull request.

File: denoiser/models/demucs_encoder.py (hunk 5 above)
Comment (mosem) [reply denoiser/models/demucs_encoder.py:139]: Not sure I understand this, can you explain this to me when we meet?

File: conf/experiment/demucs_skipless_1.yaml (hunk 1 above)
Comment (mosem) [reply conf/experiment/demucs_skipless_1.yaml:1]: done.

File: denoiser/models/autoencoder_composer.py (hunk 3 above)
Comment (mosem) [reply denoiser/models/autoencoder_composer.py:30]: done.

File: test_denoiser.py (hunk 7 above)
Comment (mosem) [reply test_denoiser.py:31]: done.

File: conf/experiment/demucs_skipless_adversarial_1.yaml (hunk 2 above)
Comment (mosem) [reply conf/experiment/demucs_skipless_adversarial_1.yaml:2]: done.

File: denoiser/models/demucs_encoder.py (hunk 5 above)
Comment (mosem) [reply denoiser/models/demucs_encoder.py]: done. implemented in autoencoder_decomposer.py

**Comment by mosem** [comment]
Are the new changes satisfactory enough to approve?

**Pull request merged**